File: android/app/src/main/java/com/innocorelabs/verbale/pipeline/ReviewRule.kt
package com.innocorelabs.verbale.pipeline

import com.innocorelabs.verbale.data.AudioDb

/**
 * What the model could not settle — the founder's answer to "what enters the queue".
 *
 * A confidently typed item with an owner and no date trouble is shown with its label and never
 * asked about. Everything else is one card. A person's confirmed or rejected is theirs: the rule
 * never overwrites it, and the reconciler's own needs_review is left as it is. Mirrored in
 * src/pipeline/reviewRule.ts against cpp/tests/golden/review_rule.json.
 */
object ReviewRule {
  /** Why the item is in the queue, in priority order — the card shows the first. */
  enum class Reason { TYPE, STATUS, OWNER, DATE }

  data class Decision(val review: String, val reason: String?)

  fun decide(
    kind: String, type: String, status: String, confidence: String, ownerKind: String,
    dateSaid: String?, dateNorm: Long?, currentReview: String,
  ): Decision {
    if (currentReview in AudioDb.Review.BY_A_PERSON) return Decision(currentReview, null)
    if (currentReview == AudioDb.Review.NEEDS_REVIEW) return Decision(currentReview, null)
    val reason = when {
      type == "uncertain" || confidence == "low" -> Reason.TYPE
      status != "open" -> Reason.STATUS
      kind == "action" && ownerKind == "unassigned" -> Reason.OWNER
      !dateSaid.isNullOrEmpty() && dateNorm == null -> Reason.DATE
      else -> null
    }
    return if (reason == null) Decision(AudioDb.Review.SUGGESTED, null)
    else Decision(AudioDb.Review.NEEDS_REVIEW, reason.name.lowercase())
  }
}

File: src/pipeline/reviewRule.ts
/**
 * What the model could not settle — the mirror of ReviewRule.kt, against
 * cpp/tests/golden/review_rule.json. The phone decides `review` with the Kotlin one when it
 * classifies; the review card uses this one to say WHY an item is in front of the person.
 */
export type ReviewReason = 'type' | 'status' | 'owner' | 'date';

export interface ReviewInput {
  kind: string;
  type: string | null;
  status: string | null;
  confidence: string | null;
  ownerKind: string;
  dateSaid: string | null;
  dateNorm: number | null;
  currentReview: string;
}

export function decideReview(i: ReviewInput): { review: string; reason: ReviewReason | null } {
  if (i.currentReview === 'confirmed' || i.currentReview === 'rejected') return { review: i.currentReview, reason: null };
  if (i.currentReview === 'needs_review') return { review: i.currentReview, reason: null };
  const reason = reviewReason(i);
  return reason ? { review: 'needs_review', reason } : { review: 'suggested', reason: null };
}

/** The first trigger in priority order, ignoring the current review — what a card leads with. */
export function reviewReason(i: ReviewInput): ReviewReason | null {
  if (i.type === 'uncertain' || i.confidence === 'low') return 'type';
  if (i.status && i.status !== 'open') return 'status';
  if (i.kind === 'action' && i.ownerKind === 'unassigned') return 'owner';
  if (i.dateSaid && i.dateNorm === null) return 'date';
  return null;
}

export const REASON_TEXT: Record<ReviewReason, string> = {
  type: 'The model was not sure what this is.',
  status: 'A later turn pushed back on this.',
  owner: 'No owner.',
  date: 'The date could not be pinned to a day.',
};
